crates/ruff/src/rules/flake8_pyi/helpers.rs

```diff
@@ -0,0 +1,54 @@
+use ruff_python_semantic::SemanticModel;
+use rustpython_parser::ast::{self, Expr, Operator};
+
+/// Traverse a "union" type annotation, applying `func` to each union member.
+/// Supports traversal of `Union` and `|` union expressions.
+/// The function is called with each expression in the union (excluding declarations of nested unions)
+/// and the parent expression (if any).
+pub(crate) fn traverse_union<'a, F>(
```

Comment:
Nit: this could be `pub(super)`, since only modules in `flake8_pyi` need to have access to it.